main.c

```diff
@@ -414,41 +412,89 @@ void animationCallback(GPTDriver* _driver) {
   (void)_driver;
 
   // If the foreground is set we skip the animation as a way to avoid it overrides the foreground
-  if(!is_foregroundColor_set) {
+  if(!is_foregroundColor_set && profiles[currentProfile].animationSpeed[currentSpeed] > 0) {
     profiles[currentProfile].callback(ledColors, ledIntensity);
   }
 }
 
 inline void sPWM(uint8_t cycle, uint8_t currentCount, uint8_t start, ioline_t port) {
-  if (start+cycle>0xFF) start = 0xFF - cycle;
+  if (start > 0xFF - cycle || cycle > 0xFF - start) {
+    start = 0xFF - cycle;
+  }
   if (start <= currentCount && currentCount < start+cycle)
     palSetLine(port);
   else
     palClearLine(port);
 }
 
-void columnCallback(GPTDriver* _driver) {
+uint16_t animationSkipCounter = 0;
+uint16_t serialCheckCounter = 0;
+
+void mainLoop(GPTDriver* _driver) {
   (void)_driver;
-  palClearLine(ledColumns[currentColumn]);
-  currentColumn = (currentColumn+1) % NUM_COLUMN;
-  if (columnPWMCount < 255) {
-    for (size_t row = 0; row < NUM_ROW; row++) {
-      const size_t ledIndex = currentColumn + (NUM_COLUMN * row);
-      const led_t keyLED = ledColors[ledIndex];
-      const uint8_t ledMask = ledMasks[ledIndex];
-      const uint8_t red = keyLED.red & ledMask;
-      const uint8_t green = keyLED.green & ledMask;
-      const uint8_t blue = keyLED.blue & ledMask;
-
-      sPWM(red, columnPWMCount, 0, ledRows[row << 2]);
-      sPWM(green, columnPWMCount, red, ledRows[(row << 2) | 1]);
-      sPWM(blue, columnPWMCount, red+green, ledRows[(row << 2) | 2]);
+
+  serialCheckCounter++;
+  /*
```

Comment:
This is okay with me. But just letting you know that from my perspective, this is not a huge speedup. Since if you track down how osalSysLock works, it is a static inline function that calls chSysLock, which in our case is implemented in the `arm_v6m.h` file. The implementation is simply __disable_irq(). And this is a compiler intrinsic that translates to a single instruction (CPSID i). It is a single cycle instruction. The same basically applies to osalSysUnlock.